### PR TITLE
XDP support for AIE Reload Reset feature on Edge platform.

### PR DESCRIPTION
**Pull request by vipangul**
#### Problem solved by the commit
This is planned feature support of XDP plugins for AIE reload reset tracked via Vitis-8456

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Planned enhancement

#### How problem was solved, alternative solutions (if any) and why they were rejected
This feature considers a new concept of config , where each config could contain multiple xclbins
that are needed for the successful run. E.g. PL only and AIE only xclbins.

#### Risks (if any) associated the changes in the commit
Moderate.

#### What has been tested and how, request additional testing if necessary
Verified on edge and client.

### CODE REVIEW

File: src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp [hunk 1]
```diff
@@ -21,26 +21,77 @@
...
+    this->hasMemoryAIM = other.hasMemoryAIM ;
+
+    for(auto &cu : other.cus)
+      this->cus[cu.first] = new ComputeUnitInstance(*cu.second) ;
+
+    for (auto &mi : other.memoryInfo)
+      this->memoryInfo[mi.first] = new Memory(*mi.second) ;
+    
+    this->ams.reserve(other.ams.size());
+    for (auto& am : other.ams) {
```
Comment (hackwa) [inline src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp:52]: Brackets aren't needed for one line

File: src/runtime_src/xdp/profile/database/static_info/device_info.cpp [hunk 2]
```diff
@@ -39,260 +38,316 @@ namespace xdp {
...
+    else
+    {
+      xclbin->aie.valid = false ;
+      missingXclbin = createXclbinFromLastConfig(XCLBIN_AIE_ONLY);
+    }
+
+    // If missing part of XclbinInfo is available. 
+    if(missingXclbin)
+    {
+      config->currentXclbins.back()->aie.numTracePLIO = loadedConfigInfos.size() == 0 ? 0 : loadedConfigInfos.back()->currentXclbins.back()->aie.numTracePLIO;
```
Comment (hackwa) [inline src/runtime_src/xdp/profile/database/static_info/device_info.cpp:149]: Is this safe if we load pl xclbin first? Why do we have to set this?

File: src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp [hunk 3]
```diff
@@ -94,17 +145,568 @@ namespace xdp {
...
+      if(bin == xclbin)
+        return true;
+    }
+    return false;
+  }
+
+    bool ConfigInfo::hasFloatingAIMWithTrace(XclbinInfo* xclbin)
+    {
+      for(auto bin : currentXclbins)
+        if(bin == xclbin)
```
Comment (hackwa) [inline src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp:326]: Do we need a check here to confirm that the xclbin in arguments is pl xclbin? This comment is for all the functions

File: src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp [hunk 4]
```diff
@@ -94,17 +145,568 @@ namespace xdp {
...
+          break ;
+        }
+      }
+
+      if(!xclbin)
+        return;
+
+      switch (moduleType)
+      {
+        case 0:
```
Comment (hackwa) [inline src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp:596]: Is there an enumeration for this?

**Review by hackwa — COMMENTED** [review]
Added some comments minor comments.

File: src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp [hunk 5]
```diff
@@ -94,17 +145,568 @@ namespace xdp {
     }
   }
 
+  AIEInfo& AIEInfo::operator=(const AIEInfo& other)
```
Comment (IshitaGhosh) [inline src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp:148]: Using "source"/"src" as the name of the source AIEInfo, in place of "other" is more intuitive and readable.

File: src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp [hunk 6]
```diff
@@ -21,26 +21,77 @@
 #include "xdp/profile/database/static_info/pl_constructs.h"
 #include "xdp/profile/database/static_info/xclbin_info.h"
 #include "xdp/profile/device/device_intf.h"
+#include "core/common/message.h"
 
 namespace xdp {
 
+  PLInfo& PLInfo::operator=(const PLInfo& other)
```
Comment (IshitaGhosh) [inline src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp:28]: Using "source"/"src" as the name of the source PLInfo, in place of "other" is more intuitive and readable.

File: build/petalinux.build [hunk 7]
```diff
@@ -1,3 +1,4 @@
 # When updating Petalinux build please file a SH ticket to retain the build
 # https://jira.xilinx.com/secure/CreateIssue!default.jspa
-PETALINUX="/proj/petalinux/2024.2/petalinux-v2024.2_05300037/tool/petalinux-v2024.2-final"
+#PETALINUX="/proj/petalinux/2024.1/petalinux-v2024.1_03142213/tool/petalinux-v2024.1-final"
+PETALINUX="/proj/petalinux/2024.1/petalinux-v2024.1_05202009/tool/petalinux-v2024.1-final"
```
Comment (IshitaGhosh) [inline build/petalinux.build:4]: Please remove this change from this PR.

File: src/runtime_src/xdp/profile/database/static_info/device_info.cpp (hunk 2 above)
Comment (vipangul) [reply src/runtime_src/xdp/profile/database/static_info/device_info.cpp:131]: Somehow github doesn't show responses I had posted to review. Re-posting
"As part of initializing profile monitors,and hence ts2mm initialization from PL Xclbin, number of trace PLIO gets recorded in AIEInfo. Hence to account that information, this check is added.
Ideally we can move number of trace PLIO information to config, I can take care of that in follow-up commit."

File: src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp (hunk 4 above)
Comment (vipangul) [reply src/runtime_src/xdp/profile/database/static_info/xclbin_info.cpp:608]: addressed via https://github.com/Xilinx/XRT/pull/8244

File: src/runtime_src/xdp/profile/database/static_info_database.cpp [hunk 8]
```diff
@@ -1329,37 +1480,44 @@ namespace xdp {
...
     if(itr != deviceInfo.end()) {
       DeviceInfo *devInfo = itr->second.get();
-      // Are we attempting to load the same xclbin multiple times?
-      XclbinInfo* xclbin = devInfo->currentXclbin() ;
-      if (xclbin && device->get_xclbin_uuid() == xclbin->uuid) {
-        return false ;
+      ConfigInfo* config = devInfo->currentConfig() ;
+      auto xclbinUuids = config->getConfigUuids() ;
+
+      if(!xclbinUuids.empty() && std::find(xclbinUuids.begin(), xclbinUuids.end(), new_xclbin_uuid) != xclbinUuids.end()) {
```
[the remark this answers is not in the document]
Comment (vipangul) [reply src/runtime_src/xdp/profile/database/static_info_database.cpp]: Thanks, addressed.

**Review by jvillarre — APPROVED** [review]

**Pull request merged**